packages/shorebird_cli/lib/src/code_push_client_wrapper.dart

```diff
@@ -0,0 +1,282 @@
+import 'package:collection/collection.dart';
+import 'package:mason_logger/mason_logger.dart';
+import 'package:shorebird_cli/src/shorebird_build_mixin.dart';
+import 'package:shorebird_cli/src/third_party/flutter_tools/lib/flutter_tools.dart';
+import 'package:shorebird_code_push_client/shorebird_code_push_client.dart';
+
+/// {@template patch_artifact_bundle}
+/// Metadata about a patch artifact that we are about to upload.
+/// {@endtemplate}
+class PatchArtifactBundle {
+  /// {@macro patch_artifact_bundle}
+  const PatchArtifactBundle({
+    required this.arch,
+    required this.path,
+    required this.hash,
+    required this.size,
+  });
+
+  /// The corresponding architecture.
+  final String arch;
+
+  /// The path to the artifact.
+  final String path;
+
+  /// The artifact hash.
+  final String hash;
+
+  /// The size in bytes of the artifact.
+  final int size;
+}
+
+/// {@template code_push_client_wrapper}
+/// Wraps [CodePushClient] interaction with logging and error handling to
+/// reduce the amount of command and command test code.
+/// {@endtemplate}
+class CodePushClientWrapper {
+  /// {@macro code_push_client_wrapper}
+  CodePushClientWrapper({
+    required this.codePushClient,
+    required this.logger,
+  });
+
+  final CodePushClient codePushClient;
+  final Logger logger;
+
+  Future<AppMetadata> getApp({required String appId}) async {
+    final app = await maybeGetApp(appId: appId);
+    if (app == null) {
+      logger.err(
+        '''
+Could not find app with id: "$appId".
+Did you forget to run "shorebird init"?''',
+      );
+      exit(ExitCode.software.code);
+    }
+
+    return app;
+  }
+
+  Future<AppMetadata?> maybeGetApp({required String appId}) async {
+    final fetchAppsProgress = logger.progress('Fetching apps');
+    try {
+      final apps = (await codePushClient.getApps()).toList();
+      fetchAppsProgress.complete();
+      return apps.firstWhereOrNull((a) => a.appId == appId);
+    } catch (error) {
+      fetchAppsProgress.fail('$error');
+      exit(ExitCode.software.code);
+    }
+  }
+
+  Future<Channel?> maybeGetChannel({
+    required String appId,
+    required String name,
+  }) async {
+    final fetchChannelsProgress = logger.progress('Fetching channels');
+    try {
+      final channels = await codePushClient.getChannels(appId: appId);
+      final channel = channels.firstWhereOrNull(
+        (channel) => channel.name == name,
+      );
+      fetchChannelsProgress.complete();
+      return channel;
+    } catch (error) {
+      fetchChannelsProgress.fail('$error');
+      exit(ExitCode.software.code);
+    }
+  }
+
+  Future<Channel> createChannel({
```

Comment:
thoughts on making the methods like this one private since they are only used internally?

Reply:
Ah good call, I didn't realize how much the `publishPatch` was hiding. I made this `visibleForTesting` to keep the test file neatly organized.